crates/bevy_ecs/src/schedule/stage.rs

```diff
@@ -1,232 +1,1126 @@
+            )
+        }
+    }
+    sorted
 }
 
-impl<S: System<In = (), Out = ()>> From<S> for SystemStage {
-    fn from(system: S) -> Self {
-        SystemStage::single(system)
+/// Rearranges given vector of system containers to match the order of the given indices.
+fn rearrange_to_order(systems: &mut Vec<impl SystemContainer>, order: &[usize]) {
+    let mut temp = systems.drain(..).map(Some).collect::<Vec<_>>();
+    for index in order {
+        systems.push(temp[*index].take().unwrap());
     }
 }
 
-pub struct RunOnce {
-    ran: bool,
-    system_id: SystemId,
-    resource_access: TypeAccess<TypeId>,
-    archetype_access: TypeAccess<ArchetypeComponent>,
+/// Populates the map of all ascendants and descendants using the given graph.
+fn populate_relations(
+    graph: &HashMap<usize, Vec<usize>>,
+    all_relations: &mut HashMap<usize, FixedBitSet>,
+) {
+    fn add_relations(
+        index: usize,
+        current_descendant: usize,
+        graph: &HashMap<usize, Vec<usize>>,
+        all_relations: &mut HashMap<usize, FixedBitSet>,
+    ) {
+        for &descendant in graph.get(&current_descendant).unwrap() {
+            all_relations.get_mut(&index).unwrap().insert(descendant);
+            all_relations
+                .entry(descendant)
+                .or_insert_with(|| FixedBitSet::with_capacity(graph.len()))
+                .insert(index);
+            add_relations(index, descendant, graph, all_relations);
+        }
+    }
+    all_relations.reserve(graph.len());
+    for index in 0..graph.len() {
+        all_relations
+            .entry(index)
+            .or_insert_with(|| FixedBitSet::with_capacity(graph.len()))
+            .insert(index);
+        add_relations(index, index, graph, all_relations);
+    }
 }
 
-impl Default for RunOnce {
-    fn default() -> Self {
-        Self {
-            ran: false,
-            system_id: SystemId::new(),
-            resource_access: Default::default(),
-            archetype_access: Default::default(),
+impl Stage for SystemStage {
+    fn run(&mut self, world: &mut World, resources: &mut Resources) {
+        // Evaluate sets' run criteria, initialize sets as needed, detect if any sets were changed.
+        let mut has_work = false;
+        for system_set in self.system_sets.iter_mut() {
+            let result = system_set.run_criteria.should_run(world, resources);
+            match result {
+                Yes | YesAndCheckAgain => has_work = true,
+                No | NoButCheckAgain => (),
+            }
+            system_set.should_run = result;
+        }
+
+        if self.systems_modified {
+            self.initialize_systems(world, resources);
+            let mut ambiguities = self.rebuild_orders_and_dependencies();
+            self.systems_modified = false;
+            self.executor.rebuild_cached_data(&mut self.parallel, world);
+            self.executor_modified = false;
+            if !ambiguities.is_empty() {
+                info!("{}", {
```

Comment:
Given how noisy this is, I think we should make it opt-in until we have resolved the warnings in all of the official bevy crates. Then we can try enabling it by default to see how it goes. As it stands it won't help us determine "will user code be too noisy", because I know for a fact that right now _every_ user will think its too noisy :smile:

Reply:
I agree here. We will need explicit ambiguities for that (and to make this useful for users IMO). That can easily be another PR on top of this though.

Reply:
We don't need explicit ambiguities to make this opt-in. We just need to re-plumb it as a plugin or a setting (filtering in the logging plugin?), or feature-gate it.

Reply:
Yes to both, please: make it optional (setting or feature-gate), and wait until everything internal to bevy is resolved/refactored, before flipping it from opt-in to opt-out.

Reply:
Added a `struct ReportExecutionOrderAmbiguities;` in ec845c5946273c984ff444498daa4750af84ad17. I don't think we need anything smarter right now.